controller/src/main/java/io/pravega/controller/store/client/StoreClientFactory.java

```diff
@@ -115,14 +121,51 @@ public ZooKeeper newZooKeeper(String connectString, int sessionTimeout, Watcher
                 this.client = new ZooKeeper(connectString, sessionTimeout, watcher, canBeReadOnly);
                 return this.client;
             } else {
-                Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
-                Preconditions.checkArgument(this.sessionTimeout == sessionTimeout, "sessionTimeout differs");
-                Preconditions.checkArgument(this.canBeReadOnly == canBeReadOnly, "canBeReadOnly differs");
-                Preconditions.checkNotNull(watcher, "watcher");
-                this.client.register(watcher);
+                try {
+                    tryInjectFailure();
+                    Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
+                    Preconditions.checkArgument(this.sessionTimeout == sessionTimeout, "sessionTimeout differs");
+                    Preconditions.checkArgument(this.canBeReadOnly == canBeReadOnly, "canBeReadOnly differs");
+                    Preconditions.checkNotNull(watcher, "watcher");
+                    this.client.register(watcher);
+                } catch (IllegalArgumentException | NullPointerException e) {
+                    if (watcher != null) {
+                        log.warn("Input argument for new ZooKeeper client ({}, {}, {}) changed with respect to existing client ({}, {}, {}).",
+                            connectString, sessionTimeout, canBeReadOnly, this.connectString, this.sessionTimeout, this.canBeReadOnly, e);
+                    } else {
+                        log.warn("Watcher for ZooKeeper client is null.", e);
+                    }
+                    closeClient(client);
+                }
                 return this.client;
             }
         }
+
+        private void closeClient(ZooKeeper client) {
+            try {
+                client.close();
+            } catch (Exception e) {
+                // We prevent throwing uncontrolled exceptions here, which may lead Curator to retry indefinitely.
+                log.error("Problems attempting to close ZooKeeper client.", e);
+            }
+        }
+
+        /**
+         * Simulates an error related to a change in ZooKeeper client parameters. The error is injected only once and
+         * this class automatically sets this flag to false when the exception is thrown.
+         */
+        @VisibleForTesting
+        @Synchronized
```

Comment:
Also, you may want to use an atomic boolean rather than synchronize the method.

Reply:
Done.